codegen/builtin_fs.go

```diff
@@ -787,6 +817,75 @@ func (dp DockerPush) Call(ctx context.Context, cln *client.Client, val Value, op
 	return NewValue(ctx, fs)
 }
 
+// hasNonStargzLayer traverses a manifest by resolving the reference and
+// walking over its children to see if there's any non-stargz layers.
+//
+// If ref points to a manifest list, the platform matcher is used to only
+// consider the current platform. Although BuildKit supports building manifest
+// lists, and multi-platform conversion is technically possible client side,
+// it requires pulling blobs locally (not via BuildKit) which is undesirable.
+//
+// See: https://github.com/containerd/stargz-snapshotter/blob/v0.6.4/nativeconverter/estargz/estargz.go
+func hasNonStargzLayer(ctx context.Context, resolver remotes.Resolver, platform platforms.MatchComparer, ref string) (bool, error) {
```

Comment:
Can we add a test for this?

Reply:
Okay, and maybe this makes sense to move to this to under `pkg/`